### PR TITLE
Use dbiterator, db::insert and db::update

### PR DESCRIPTION
**Pull request by trasher**

### CODE REVIEW

File: inc/deploytaskjob.class.php [hunk 1]
```diff
@@ -166,21 +185,24 @@ function saveDatas($params) {
...
-            '$tasks_id', 'job_".$tasks_id."_$i', NOW(), '0',
-            '$plugins_id', '".$task['method']."', '$definition', '$action',
-            '".$task['retry_nb']."', '".$task['retry_time']."', 'minutes', '0'
-         )";
-         $i++;
-      }
-      foreach ($sql_tasks as $query) {
-         $DB->query($query);
+         $stmt->bind_param(
+            'ssssssssssss',
```
Comment (orthagh) [inline inc/deploytaskjob.class.php]: ?

File: inc/networkportlog.class.php [hunk 2]
```diff
@@ -258,24 +271,27 @@ static function cronCleannetworkportlogs() {
...
-            $query_delete = "DELETE FROM `".$pfNetworkPortLog->getTable()."`
-               WHERE `plugin_fusioninventory_mappings_id`='".
-                    $data['plugin_fusioninventory_mappings_id']."' ";
-
             switch ($data['days']) {
-
                case '-1':
-                  $DB->query($query_delete);
+                  $DB->delete(
+                     $pfNetworkPortLog->getTabNameForItem(), [
```
Comment (orthagh) [inline inc/networkportlog.class.php]: getTabNameForItem -> getTable

File: inc/networkportlog.class.php [hunk 3]
```diff
@@ -258,24 +271,27 @@ static function cronCleannetworkportlogs() {
...
 
                case '0': // never delete
                   break;
 
                default:
-                  $query_delete .= " AND `date_mod` < date_add(now(), interval -".
-                                       $data['days']." day)";
-                  $DB->query($query_delete);
+                  $DB->delete(
+                     $pfNetworkPortLog->getTabNameForItem(), [
```
Comment (orthagh) [inline inc/networkportlog.class.php]: getTabNameForItem -> getTable

File: inc/networkporttype.class.php [hunk 4]
```diff
@@ -301,19 +301,46 @@ function init() {
...
-      if ($DB->numrows($result) > 0) {
+      $iterator = $DB->request([
+         'FROM'   => 'glpi_plugin_fusioninventory_networkporttypes',
+         'WHERE'  => ['import' => 1]
+      ]);
+      if (count($iterator) > 0) {
          $install = 0;
       }
 
+      $it = $it = new DBmysqlIterator($DB);
```
Comment (orthagh) [inline inc/networkporttype.class.php]: double assignation

File: install/update.php [hunk 5]
```diff
@@ -889,12 +956,29 @@ function pluginFusioninventoryUpdate($current_version, $migrationname = 'Migrati
...
+      'FROM'   => 'glpi_computers',
+      'WHERE'  => ['entities_id' => ['>', 0]]
+   ]);
+   if (count($iterator)) {
+      $update = $DB->buildUpdate(
+         'glpi_computers_softwareversions', [
+            'entities_id'  => new \QueryParam()
+         ], [
+            'computers_id' => new \QueryParam(),
+            'is_dynamic'   => 1
```
Comment (orthagh) [inline install/update.php]: it seems you miss a param here: entities_id => 0.
Correctly set in first param of ``bind_param`` function below.

File: install/update.php [hunk 6]
```diff
@@ -1299,39 +1442,36 @@ function do_config_migration($migration) {
...
-         WHERE `type`='version'
-         LIMIT 1, 10";
-      $result=$DB->query($query);
-      while ($data=$DB->fetch_array($result)) {
-         $DB->query("DELETE FROM `glpi_plugin_fusioninventory_configs`
-            WHERE `".$id."`='".$data[$id]."'");
+      $iterator = $DB->request([
+         'FROM'   => 'glpi_plugin_fusioninventory_configs',
+         'WHERE'  => ['type' => 'version'],
+         'START'  => 0,
```
Comment (orthagh) [inline install/update.php]: the offset seems different (0 != 1)

File: install/update.php [hunk 7]
```diff
@@ -1299,39 +1442,36 @@ function do_config_migration($migration) {
...
+         'LIMIT'  => 10
+      ]);
+      while ($data = $iterator->next()) {
+         $DB->delete(
```
Comment (orthagh) [inline install/update.php]: prepared statement ?
or maybe a single ``delete`` for the two statements ``SELECT`` and ``DELETE``
Actually raw mysql statements supports ``LIMIT`` keyword, did the iterator do ?

File: install/update.php [hunk 8]
```diff
@@ -2607,15 +2849,19 @@ function do_biosascomponentmigration() {
...
-               $existing = $DB->fetch_assoc($result);
+            $iterator = $DB->request([
+               'SELECT' => 'id',
+               'FROM'   => 'glpi_devicefirmwares',
+               'WHERE'  => [
+                  'date'               => $data['bios_date'],
+                  'version'            => $data['bios_version'],
+                  'manufacturers_id'   => $data['bios_manufacturers_id']
+               ],
+               'START'  => 0,
```
Comment (orthagh) [inline install/update.php]: START/OFFSET differs from LIMIT/OFFSET.
And OFFSET exists as crit in iterator ?

File: install/update.php [hunk 9]
```diff
@@ -2884,210 +3133,210 @@ function do_networkport_migration($migration) {
...
+   /*
+    * Table glpi_plugin_fusioninventory_networkportlogs
+    */
+   $newTable = "glpi_plugin_fusioninventory_networkportlogs";
+   if ($DB->tableExists("glpi_plugin_tracker_snmp_history")) {
+      // **** Update history
+      update213to220_ConvertField($migration);
+
+      // **** Migration network history connections
+      $query = "SELECT count(ID) FROM `glpi_plugin_tracker_snmp_history`
```
Comment (orthagh) [inline install/update.php]: did you miss this query ?

File: inc/networkportlog.class.php (hunk 2 above)
Comment (wawax) [inline inc/networkportlog.class.php]: $pfNetworkPortLog->getTabNameForItem() => $pfNetworkPortLog->getTable()

File: inc/networkportlog.class.php (hunk 3 above)
Comment (wawax) [inline inc/networkportlog.class.php]: $pfNetworkPortLog->getTabNameForItem() => getTable

File: install/update.php (hunk 6 above)
Comment (trasher) [reply install/update.php]: Indeed

File: install/update.php (hunk 7 above)
Comment (trasher) [reply install/update.php]: DBIterator does support LIMIt: http://glpi-developer-documentation.readthedocs.io/en/master/devapi/database/dbiterator.html#request-pager; but not DB::update() nor DB::delete().
I've changed with a prepared statement

File: install/update.php (hunk 8 above)
Comment (trasher) [reply install/update.php]: OFFSET does not exists; I should write LIMIT...

File: install/update.php (hunk 9 above)
Comment (trasher) [reply install/update.php]: Yes... And no :p I've not replace all SELECT queries yet; espacially the more complicated. This one is not; and should be migrated.

File: inc/networkportlog.class.php (hunk 2 above)
Comment (trasher) [reply inc/networkportlog.class.php]: Both fixed; thanks

**Comment by trasher** [comment]
I've added a few new iterator replacements (in a separate commit to make review easier); but I'll stop for now; I've already spend a lot of time on this PR ;)

**Pull request merged**